### PR TITLE
Update for recent Julia and Package versions

### PR DESCRIPTION
**Pull request by alecloudenback**
Wanted to run on recent Julia version (e.g. Julia 1.7) and got a ton of package downgrades. Started to update for newer dependency versions but have run into the following issues:

First: 
```
Error During Test at /Users/alec/.julia/dev/Miletus/test/montecarlo.jl:31
  Test threw exception
  Expression: isapprox(value(m, c), value(mcm, c), rtol = 0.01)
  TypeError: in CurrencyQuantity, in T, expected T<:Real, got Type{Number}
  Stacktrace:
    [1] promote_type
      @ ~/.julia/dev/Miletus/src/currency.jl:88 [inlined]
    [2] _mean_promote(x::CurrencyQuantity{CurrencyUnit{:USD}, Float64}, y::CurrencyQuantity{CurrencyUnit{:USD}, Float64})
      @ Statistics /Users/sabae/src/julia/usr/share/julia/stdlib/v1.7/Statistics/src/Statistics.jl:166
    [3] mean(f::typeof(identity), itr::Base.Generator{Miletus.ScenarioIterator{MonteCarloModel{CoreModel{CurrencyQuantity{CurrencyUnit{:USD}, Float64}, Miletus.TermStructure.ConstantContinuousYieldCurve{Actual365, Float64}, Miletus.TermStructure.ConstantContinuousYieldCurve{Actual365, Float64}}, StepRange{Date, Day}, CurrencyQuantity{CurrencyUnit{:USD}, Float64}}}, Miletus.var"#5#6"{Miletus.When{Miletus.LiftObs{typeof(==), Tuple{Miletus.DateObs, Miletus.ConstObs{Date}}, Bool}, SingleStock}, Int64}})
      @ Statistics /Users/sabae/src/julia/usr/share/julia/stdlib/v1.7/Statistics/src/Statistics.jl:74
    [4] mean
      @ /Users/sabae/src/julia/usr/share/julia/stdlib/v1.7/Statistics/src/Statistics.jl:44 [inlined]
    [5] value(m::MonteCarloModel{CoreModel{CurrencyQuantity{CurrencyUnit{:USD}, Float64}, Miletus.TermStructure.ConstantContinuousYieldCurve{Actual365, Float64}, Miletus.TermStructure.ConstantContinuousYieldCurve{Actual365, Float64}}, StepRange{Date, Day}, CurrencyQuantity{CurrencyUnit{:USD}, Float64}}, c::Miletus.When{Miletus.LiftObs{typeof(==), Tuple{Miletus.DateObs, Miletus.ConstObs{Date}}, Bool}, SingleStock})
      @ Miletus ~/.julia/dev/Miletus/src/models/montecarlo.jl:94
    [6] top-level scope
      @ /Users/sabae/src/julia/usr/share/julia/stdlib/v1.7/Test/src/Test.jl:445
    [7] include(fname::String)
      @ Base.MainInclude ./client.jl:451
    [8] top-level scope
      @ ~/.julia/dev/Miletus/test/runtests.jl:11
    [9] include(fname::String)
      @ Base.MainInclude ./client.jl:451
   [10] top-level scope
      @ none:6
   [11] eval
      @ ./boot.jl:373 [inlined]
   [12] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:268
   [13] _start()
      @ Base ./client.jl:495
ERROR: LoadError: There was an error during testing
in expression starting at /Users/alec/.julia/dev/Miletus/test/montecarlo.jl:31
in expression starting at /Users/alec/.julia/dev/Miletus/test/runtests.jl:11
ERROR: Package Miletus errored during testing
```

Minimal reproducing case for this is:

```julia
mean([100USD,100USD])
```

Statistics doesn't like the currency to be a `Number`, so I tried editing line 11 in currency.jl to be:

```julia
struct CurrencyQuantity{U<:CurrencyUnit,T<:Real} <: Real
```

Then there's a method ambiguity issue with ForwardDiff:

```
Error During Test at /Users/alec/.julia/dev/Miletus/test/greeks.jl:20
  Test threw exception
  Expression: delta(gbm1, c1; autodiff = true) ≈ delta(gbm1, c1)
  MethodError: *(::CurrencyQuantity{CurrencyUnit{:USD}, ForwardDiff.Dual{Nothing, Float64, 1}}, ::ForwardDiff.Dual{Nothing, Float64, 1}) is ambiguous. Candidates:
    *(x::Real, y::ForwardDiff.Dual{Ty}) where Ty in ForwardDiff at /Users/alec/.julia/packages/ForwardDiff/5gUap/src/dual.jl:140
    *(x::CurrencyQuantity{U}, y::Real) where U<:CurrencyUnit in Miletus.Currency at /Users/alec/.julia/dev/Miletus/src/currency.jl:46
  Possible fix, define
    *(::CurrencyQuantity{U}, ::ForwardDiff.Dual{Ty}) where {U<:CurrencyUnit, Ty}
  Stacktrace:
   [1] black(F::CurrencyQuantity{CurrencyUnit{:USD}, ForwardDiff.Dual{Nothing, Float64, 1}}, K::CurrencyQuantity{CurrencyUnit{:USD}, Float64}, s::Float64)
     @ Miletus ~/.julia/dev/Miletus/src/utils/black.jl:47
   [2] value(m::GeomBMModel{CoreModel{CurrencyQuantity{CurrencyUnit{:USD}, ForwardDiff.Dual{Nothing, Float64, 1}}, Miletus.TermStructure.ConstantContinuousYieldCurve{Actual365, Float64}, Miletus.TermStructure.ConstantContinuousYieldCurve{Actual365, Float64}}, Float64}, c::EuropeanCall{SingleStock, CurrencyQuantity{CurrencyUnit{:USD}, Float64}})
     @ Miletus ~/.julia/dev/Miletus/src/models/geombm.jl:54
   [3] delta(::GeomBMModel{CoreModel{CurrencyQuantity{CurrencyUnit{:USD}, Float64}, Miletus.TermStructure.ConstantContinuousYieldCurve{Actual365, Float64}, Miletus.TermStructure.ConstantContinuousYieldCurve{Actual365, Float64}}, Float64}, ::EuropeanCall{SingleStock, CurrencyQuantity{CurrencyUnit{:USD}, Float64}}; autodiff::Bool)
     @ Miletus ~/.julia/dev/Miletus/src/greeks.jl:51
   [4] macro expansion
     @ /Users/sabae/src/julia/usr/share/julia/stdlib/v1.7/Test/src/Test.jl:445 [inlined]
   [5] top-level scope
     @ ~/.julia/dev/Miletus/test/greeks.jl:20
ERROR: LoadError: There was an error during testing
in expression starting at /Users/alec/.julia/dev/Miletus/test/greeks.jl:3
in expression starting at /Users/alec/.julia/dev/Miletus/test/runtests.jl:12
ERROR: Package Miletus errored during testing
```
I'm unsure of the best way to proceed from here?

### CODE REVIEW

**Comment by andreasnoack** [comment]
I'm not sure if it will fix the issue but I don't think the package is supposed to overload `promote_type`. Instead, I believe we should define `promote_rule`.

**Comment by alecloudenback** [comment]
> I'm not sure if it will fix the issue but I don't think the package is supposed to overload `promote_type`. Instead, I believe we should define `promote_rule`.

`promote_type` doesn't seem to be extending (a ChainRule?) and is local to the package I think.

**Comment by alecloudenback** [comment]
I think the updates are good to be reviewed now. I'm not sure if the CI is still functional; I tried to add a github actions version but it didn't trigger any tests, then subsequently removed.  To see tests passing, it may need someone from @julpro to update the tests to something more current.

Locally:

```julia-repl

(Miletus) pkg> test
     Testing Miletus
...
Precompiling project...
  3 dependencies successfully precompiled in 3 seconds. 66 already precompiled.
     Testing Running tests...
     Testing Miletus tests passed 
```

File: Project.toml [hunk 1]
```diff
@@ -17,19 +17,11 @@ StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
...
-Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
-
-[targets]
-test = ["CSV", "DataFrames", "Test"]
+Optim = "^1"
+Reexport = "^1"
+SpecialFunctions = "^1"
+StatsBase = "0.29,0.30,0.31,0.32,^0.33"
+StatsFuns = "^0.8,^0.9"
+julia = "^1.3"
```
Comment (aviks) [inline Project.toml:27]: I'd probably just limit this to 1.6 now, since that is the LTS version.

File: test/Manifest.toml [hunk 2]
```diff
@@ -213,3 +261,19 @@ deps = ["DataAPI", "Random", "Test"]
...
+deps = ["Artifacts", "Libdl", "OpenBLAS_jll"]
+uuid = "8e850b90-86db-534c-a0d3-1478176c7d93"
+
+[[nghttp2_jll]]
+deps = ["Artifacts", "Libdl"]
+uuid = "8e850ede-7688-5339-a07c-302acd2aaf8d"
+
+[[p7zip_jll]]
+deps = ["Artifacts", "Libdl"]
+uuid = "3f19e933-33d8-53b3-aaab-bd5110c3b7a0"
```
Comment (aviks) [inline test/Manifest.toml:279]: Should this file be deleted as well?

**Review by aviks — COMMENTED** [review]
Thanks this looks pretty good. I have added a few comments, neither of them are very critical. I'll add a github actions test runner after this is merged.

**Comment by alecloudenback** [comment]
@aviks edits have been resolved (I'm not sure if you get notified when I do that)

**Comment by aviks** [comment]
Github actions tests are running after #6

**Pull request merged**